flare/mgp/map3b.py

```diff
@@ -122,98 +120,266 @@ def construct_grids(self):
-        """
-        Loop over different parts of the training set. from element s to element e
-
-        Args:
-            name: name of the gp instance
-            s: start index of the training data parition
-            e: end index of the training data parition
-            env12: AtomicEnvironment container of the triplet
-            kernel_info: return value of the get_3b_kernel
-        """
-
-        grid_kernel, _, _, cutoffs, hyps, hyps_mask = kernel_info
-
-        args = from_mask_to_args(hyps, cutoffs, hyps_mask)
-        r_cut = cutoffs['threebody']
-
-        grids = self.construct_grids()
-
-        if (e-s) == 0:
-            return np.empty((grids.shape[0], 0), dtype=np.float64)
-        coords = np.zeros((grids.shape[0], 9), dtype=np.float64) # padding 0
-        coords[:, 0] = np.ones_like(coords[:, 0])
-
-        fj, fdj = triplet_cutoff(grids, r_cut, coords, derivative=True) # TODO: add cutoff func
-        fdj = fdj[:, [0]]
-
-        if self.map_force:
-            prefix = 'force'
-        else:
-            prefix = 'energy'
-
-        if force_block:
-            training_data = _global_training_data[name]
-            kern_type = f'{prefix}_force'
-        else:
-            training_data = _global_training_structures[name]
-            kern_type = f'{prefix}_energy'
-
-        k_v = []
-        chunk_size = 32 ** 3
-        n_grids = grids.shape[0]
-        if n_grids > chunk_size:
-            n_chunk = ceil(n_grids / chunk_size)
-        else:
-            n_chunk = 1
-
-        for m_index in range(s, e):
-            data = training_data[m_index]
-            kern_vec = []
-            for g in range(n_chunk):
-                gs = chunk_size * g
-                ge = np.min((chunk_size * (g + 1), n_grids))
-                grid_chunk = grids[gs:ge, :]
-                fj_chunk = fj[gs:ge, :]
-                fdj_chunk = fdj[gs:ge, :]
-                kv_chunk = grid_kernel(kern_type, data, grid_chunk, fj_chunk, fdj_chunk,
-                                       env12.ctype, env12.etypes, *args)
-                kern_vec.append(kv_chunk)
-            kern_vec = np.hstack(kern_vec)
-            k_v.append(kern_vec)
-
-        if len(k_v) > 0:
-            k_v = np.vstack(k_v).T
-        else:
-            k_v = np.zeros((grids.shape[0], 0))
-
-        return k_v
+    if derivative:
+        for d in range(3):
+            inds = np.arange(3) * 3 + d
+            f0, df0 = cutoff_func(r_cut, triplets, coords[:, inds])
+            dfj = (
+                df0[:, 0] * f0[:, 1] * f0[:, 2]
+                + f0[:, 0] * df0[:, 1] * f0[:, 2]
+                + f0[:, 0] * f0[:, 1] * df0[:, 2]
+            )
+            dfj_list[:, d] = dfj
+    else:
+        f0, _ = cutoff_func(r_cut, triplets, 0)  # (n_grid, 3)
+
+    fj = f0[:, 0] * f0[:, 1] * f0[:, 2]  # (n_grid,)
+    fj = np.expand_dims(fj, axis=1)
+
+    return fj, dfj_list
+
+
+def get_hyps_for_kern(hyps, cutoffs, hyps_mask, c2, etypes2):
```

Comment:
Maybe it is better to move this part of code to the parameters.py. The point of having a parameter class is to collect all the relevant codes in one file, so it is easier to maintain.

Reply:
yeah, will move the `get_hyps_for_kern` in map2b and map3b into the parameter.py